File: call-in-script/4.eventinfo.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

import os
import sys
import datetime
import subprocess

os.putenv("SAC_DISPLAY_COPYRIGHT", '0')

if len(sys.argv) != 7:
    sys.exit("Usage: python %s dirname yyyy-mm-ddThh:mm:ss.xxx evlo evla evdp mag" % sys.argv[0])

dir, origin, evlo, evla, evdp, mag = sys.argv[1:]

o = datetime.datetime.strptime(origin,'%Y-%m-%dT%H:%M:%S.%f')
# 计算发震日期是一年中的第几天
jday = o.strftime("%j")
# 提取毫秒值
msec = int(o.microsecond / 1000 + 0.5)

os.chdir(dir)

p = subprocess.Popen(['sac'], stdin=subprocess.PIPE)
s = "wild echo off \n";
s += "r *.SAC \n"
s += "synchronize \n"
s += "ch o gmt %s %s %s %s %s %d \n" % (o.year, jday, o.hour, o.minute, o.second, msec)
s += "ch allt (0 - &1,o&) iztype IO \n"
s += "ch evlo %s evla %s evdp %s mag %s \n" % (evlo, evla, evdp, mag)
s += "wh \n"
s += "q \n"
p.communicate(s.encode())

os.chdir("..")

File: call-in-script/4.eventinfo.pl
#!/usr/bin/env perl
use strict;
use warnings;
use POSIX qw(strftime);

@ARGV == 6 or die
    "Usage: perl $0 dirname yyyy-mm-ddThh:mm:ss.xxx evlo evla evdp mag\n";
my ($dir, $origin, $evlo, $evla, $evdp, $mag) = @ARGV;

# 对发震时刻做解析
my ($date, $time) = split "T", $origin;
my ($year, $month, $day) = split "-", $date;
my ($hour, $minute, $second) = split ":", $time;
# 秒和毫秒均为整数
my ($sec, $msec) = split /\./, $second;
$msec = int(($second - $sec) * 1000 + 0.5);

# 计算发震日期是一年中的第几天
my $jday = strftime("%j", $second, $minute, $hour, $day, $month-1, $year-1900);

chdir $dir;
open(SAC, "| sac") or die "Error in opening SAC\n";
print SAC "wild echo off \n";
print SAC "r *.SAC \n";
print SAC "synchronize \n";   # 同步所有文件的参考时刻
print SAC "ch o gmt $year $jday $hour $minute $sec $msec \n";
print SAC "ch allt (0 - &1,o&) iztype IO \n";
print SAC "ch evlo $evlo evla $evla evdp $evdp mag $mag \n";
print SAC "wh \n";
print SAC "q \n";
close(SAC);

chdir "..";
